<a href="https://colab.research.google.com/github/DepartmentOfStatisticsPUE/cda-2021/blob/main/notebooks/cda_2021_04_27_lecture.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Introduction to GLM


Introduction to the `glm` function in R.

In [1]:
head(mtcars)

mpg  cyl disp hp  drat wt    qsec  vs am gear carb
Mazda RX4         21.0 6   160  110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag     21.0 6   160  110 3.90 2.875 17.02 0  1  4    4   
Datsun 710        22.8 4   108   93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive    21.4 6   258  110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout 18.7 8   360  175 3.15 3.440 17.02 0  0  3    2   
Valiant           18.1 6   225  105 2.76 3.460 20.22 1  0  3    1

First, let's use the `lm` function.

In [2]:
model1 <- lm(formula = mpg ~ wt + am, data = mtcars)
summary(model1)


Call:
lm(formula = mpg ~ wt + am, data = mtcars)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.5295 -2.3619 -0.1317  1.4025  6.8782 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 37.32155    3.05464  12.218 5.84e-13 ***
wt          -5.35281    0.78824  -6.791 1.87e-07 ***
am          -0.02362    1.54565  -0.015    0.988    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 3.098 on 29 degrees of freedom
Multiple R-squared:  0.7528,	Adjusted R-squared:  0.7358 
F-statistic: 44.17 on 2 and 29 DF,  p-value: 1.579e-09


Them, for the comparison we will use `glm` function. If we use gaussian(link = "identity"), we mean that:

+ we assume gaussian distribution of residuals  $N(\mu, \sigma^2)$
+ link = "identity" means that $g(\mu) = \mu$


In [5]:
model2 <- glm(formula = mpg ~ wt + am, data = mtcars, family = gaussian(link = "identity")) ## family = "gaussian"
summary(model2)


Call:
glm(formula = mpg ~ wt + am, family = gaussian(link = "identity"), 
    data = mtcars)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-4.5295  -2.3619  -0.1317   1.4025   6.8782  

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) 37.32155    3.05464  12.218 5.84e-13 ***
wt          -5.35281    0.78824  -6.791 1.87e-07 ***
am          -0.02362    1.54565  -0.015    0.988    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for gaussian family taken to be 9.597231)

    Null deviance: 1126.05  on 31  degrees of freedom
Residual deviance:  278.32  on 29  degrees of freedom
AIC: 168.03

Number of Fisher Scoring iterations: 2


In [6]:
sqrt(9.597231)

[1] 3.09794

Technical note regardin R and extracting information from R objects.

In [15]:
coef(model2) ## extract coefficients

(Intercept)          wt          am 
37.32155131 -5.35281145 -0.02361522

In [16]:
coef(summary(model2)) ## this will extract table from object created by function summary

Estimate    Std. Error t value     Pr(>|t|)    
(Intercept) 37.32155131 3.0546385  12.21799285 5.843477e-13
wt          -5.35281145 0.7882438  -6.79080719 1.867415e-07
am          -0.02361522 1.5456453  -0.01527855 9.879146e-01

In [20]:
head(model.matrix(model2))  ## this is X

(Intercept) wt    am
Mazda RX4         1           2.620 1 
Mazda RX4 Wag     1           2.875 1 
Datsun 710        1           2.320 1 
Hornet 4 Drive    1           3.215 0 
Hornet Sportabout 1           3.440 0 
Valiant           1           3.460 0

In [18]:
head(predict(model2)) ## this extracts hat(y) -- predictions from the model

Mazda RX4     Mazda RX4 Wag        Datsun 710    Hornet 4 Drive 
         23.27357          21.90860          24.87941          20.11226 
Hornet Sportabout           Valiant 
         18.90788          18.80082

In [21]:
sigma(model2) ## estimated standard error of residuals 

[1] 3.09794

In [22]:
logLik(model2) ## value of log-likelihood function

'log Lik.' -80.01459 (df=4)

## Poisson regression

Now, let's model number of gears (`gear`) in the car using their weight (`wt`). For that purpose we will use `mtcars` dataset.

In [32]:
model3 <- glm(formula = gear ~ wt, data = mtcars, family = poisson(link = "log"))
summary(model3)


Call:
glm(formula = gear ~ wt, family = poisson(link = "log"), data = mtcars)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-0.53085  -0.23403  -0.05833   0.16017   0.74850  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  1.69198    0.31842   5.314 1.07e-07 ***
wt          -0.12242    0.09812  -1.248    0.212    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 4.4634  on 31  degrees of freedom
Residual deviance: 2.8792  on 30  degrees of freedom
AIC: 108.35

Number of Fisher Scoring iterations: 4


The assumed model with the estimated parameters is as follows

$$
y = \exp(1.69 - 0.12 \times wt) 
$$

Interpretation of parameters: 

+ we cannot use the reported parameters for interpretation,
+ we need to use a transformated version using the `log()` function.

$$
log(y) = 1.69 - 0.12 \times wt
$$

if `wt` increases by one unit then `y` decreases by -0.12%

In [35]:
set.seed(123)
n <- 10000
z <- rbinom(n = n, prob = 0.7, size = 1)
theta_true <- c(1, 1.5)
lambda_true <- theta_true[1] + theta_true[2]*z ## lambda_i = 1 + 1.5*z_i
y <- rpois(n = n, lambda = lambda_true)
df <- data.frame(z, y)
head(df, 3)

z y
1 1 2
2 0 0
3 1 4

In [36]:
model3 <- glm(formula = y ~ z, data = df, family = poisson(link = "identity"))
summary(model3)


Call:
glm(formula = y ~ z, family = poisson(link = "identity"), data = df)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.2333  -1.0770   0.0322   0.3106   3.9545  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)    
(Intercept)  0.96816    0.01811   53.46   <2e-16 ***
z            1.52560    0.02611   58.43   <2e-16 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 13804  on 9999  degrees of freedom
Residual deviance: 11073  on 9998  degrees of freedom
AIC: 33271

Number of Fisher Scoring iterations: 3
